src/Compilers/Core/Portable/InternalUtilities/NoThrowStream.cs

```diff
@@ -0,0 +1,53 @@
+﻿
+using System;
+using System.Diagnostics;
+using System.IO;
+
+namespace Roslyn.Utilities
+{
+    internal class NoThrowStreamDisposer : IDisposable
+    {
+        private bool? _failed; // Nullable to assert that this is only checked after dispose
+        private readonly string _filePath;
+        private readonly TextWriter _writer;
+        private readonly Action<Exception, string, TextWriter> _exceptionHandler;
+
+        public Stream Stream { get; }
+
+        public bool Failed
+        {
+            get
+            {
+                Debug.Assert(_failed != null);
+                return _failed.GetValueOrDefault();
+            }
+        }
+        
+        public NoThrowStreamDisposer(
+            Stream stream,
+            string filePath,
+            TextWriter writer,
+            Action<Exception, string, TextWriter> exceptionHandler)
+        {
+            Stream = stream;
+            _failed = null;
+            _filePath = filePath;
+            _writer = writer;
+            _exceptionHandler = exceptionHandler;
+        }
+
+        public void Dispose()
+        {
+            try
+            {
+                Stream.Dispose();
```

Comment:
>Stream.Dispose(); [](start = 16, length = 17)

If I remember correctly, implementation should be ready to deal with multiple Dispose calls. Do we want to make sure that subsequent calls do not override initial result?